Mission service - add section on mission execution (#529)
mavlink/mavlink-devguide: en/services/mission.md

- [timeouts](#timeout) are not considered errors.
- Out-of-sequence messages in mission upload/download are recoverable, and are not treated as errors.

## Executing Missions

> Following mission upload, or vehicle restart, or mission reset, all jump counter loops are set to their initial values.

Author: Hmmm. This doc treats full mission upload as "standard" and partial as non-standard. I don't really agree with that, but it makes no difference here, so ...

```suggestion
Following full mission upload, or vehicle restart, or mission reset, all jump counter loops are set to their initial values.
```

> Unless a mission has just been uploaded that explicitly sets the current mission item, the current mission item is set to 1.

Author: @peterbarker Yes I do. Thanks for clarification. There is a further exception here that if you are flying the mission item is not reset. That allows uploading a new mission and it continuing on the same mission item.
```suggestion
The current mission item is normally reset to one on mission reset or vehicle restart, or on mission upload when the vehicle is landed. The current mission item may retain its current value if a mission is uploaded in-air or it may be set to the mission item that has `current` set in a mission item (this field is ignored in ArduPilot but respected by PX4).
```

> The mission is executed when the vehicle is armed and in the flight stack's mission mode.

Author: @peterbarker I'll change that to be a minimum requirement.

```suggestion
For a mission to be executed the vehicle must be armed and in the flight stack's mission mode.
There may (or may not) be further requirements to start a mission.
For example PX4 will start as soon as you arm in mission mode, while ArduCopter in its default configuration requires that the throttle is raised.
Similarly, planes that are configured for throttle/hand launch may require a minimum acceleration before the mission is executed. 
```

While executing, the vehicle will progress sequentially through the mission items, looping and jumping in response to jump commands.

The mission can be paused by changing to another flight mode, such as Hold/loiter, and restarted by changing back to mission mode (when armed).
On restart the vehicle will _generally_ continue towards the same waypoint item as when it was paused.
In some cases flight stacks may start from the preceding waypoint, for example, a paused camera survey may restart on the previous waypoint to ensure the whole paused leg is captured.
Note that there is no concept of "stopping" a mission independent of pausing it.

On completion of all mission items the mission will "complete".
At this point some flight stacks will loiter within the mission mode, while others will transition to some other mode (such as "Hold") and not allow you to return to mission mode.
In either case, the mission cannot be restarted unless it is [reset](#resetting-missions) (which usually happens after the vehicle lands and disarms).

While the mission is executing the current mission item can be changed via MAVLink ([MAV_CMD_DO_SET_MISSION_CURRENT](#MAV_CMD_DO_SET_MISSION_CURRENT)).

> Note however that this does not reset the jump counters, and if the mission is complete, will not restart the mission.

Reviewer: Setting the waypoint to zero in ArduPilot will reset the counters.

ArduCopter (and some other ArduPilot vehicles) reset the mission upon disarm.

Author: I'll move all the caveats into the reset the mission section so we can handle in one place.

```suggestion
Note however that unless the command [resets the mission](#resetting-missions), this does not reset the jump counters, and if the mission is complete, will not restart the mission.
```

### Starting Missions

A flight stack will usually run mission checks before allowing a vehicle to enter the mission executing mode (typically checks are run when first executing a mission, but some flight stacks might also check before continuing a paused mission).

Provided the checks pass, you can start a mission by either switching to mission mode and arming, or arming and switching to mission mode.
With mavlink we can change the mode using [MAV_CMD_DO_SET_MODE](../messages/common.md#MAV_CMD_DO_SET_MODE) to set the mode and ARM using[MAV_CMD_COMPONENT_ARM_DISARM](../messages/common.md#MAV_CMD_COMPONENT_ARM_DISARM).

If supported the [MAV_CMD_MISSION_START](../messages/common.md#MAV_CMD_MISSION_START) command will both arm the vehicle (if necessary) and switch to the mission mode.
This command also allows you to specify a particular start and end item in the mission.

> ### Pausing/Stopping Missions

Reviewer: I find it *very* strange that there's no coverage of `MAV_CMD_DO_PAUSE_CONTINUE` here!  That's the more appropriate malvink-ish way of doing this behaviour, I'd think.

Saying "do-set-mode" won't always stop the vehicle from executing waypoints - for example, ArduPlane when you've got a `DO_LAND_START`.  If you switch modes to RTL it's essentially the same thing as changing waypoints - the Plane will jump to its landing sequence.

Author: > I find it very strange that there's no coverage of MAV_CMD_DO_PAUSE_CONTINUE here! That's the more appropriate malvink-ish way of doing this behaviour, I'd think.

Doh. That's my brain fade. Thanks. Will fix.

> Saying "do-set-mode" won't always stop the vehicle from executing waypoints - for example, ArduPlane when you've got a DO_LAND_START. If you switch modes to RTL it's essentially the same thing as changing waypoints - the Plane will jump to its landing sequence.

@peterbarker This is true on PX4 too but there is a nuance that waypoints are being executed but you're still in Return mode, not mission mode. Is it the same on ArduPlane?

The TLDR here is that PX4 RTL has settings to execute a return mode landing following a mission landing pattern. So you are following waypoints, but you're not in a mission.

Reviewer: > @peterbarker This is true on PX4 too but there is a nuance that waypoints are being executed but you're still in Return mode, not mission mode. Is it the same on ArduPlane?

I believe so.  Unless behaviour has changed the vehicle will briefly appear to go into RTL mode but will then appear to go back to AUTO mode.

Author: @peterbarker What if you're not in auto mode to start with - will it go to RTL mode and still then go to AUTO mode to do the landing? That seems odd if the intent was to RTL.

Author: In any case, fixed.

Pause and stop missions by changing mode.

> In MAVLink this can be done using [MAV_CMD_DO_SET_MODE](../messages/common.md#MAV_CMD_DO_SET_MODE) (generally a vehicle will also support changing modes via an RC controller/joystick too).

Reviewer 2: E.g. to Hold, at least for PX4.

Author: Any mode change will do it. I didn't mention a specific mode because there's also some differences in Copter vs Fixed wing with related to what happens if you touch sticks in a mission (Fixed wing does not switch to Hold)

### Resetting Missions

Once a mission has started it will iterate through the mission items, jumping and looping as indicated by jump mission items, until it completes.

> While the mission is running you can change the current mission item, but this does not reset the jump counters used for loops, and once the mission is complete, changing the current mission item will not restart it.

Reviewer: As mentioned, changing to wp 0 in ArduPilot will reset.

> The mission will reset after you land and the vehicle disarms.

Reviewer 2: Are you sure?

Author: I thought so. What did you think - only on reboot?

Reviewer 2: On reboot for sure.

Reviewer 2: And on re-upload, or manual reset to 0.

Reviewer: ArduPlane will not reset counters if you land the vehicle in auto.  It's a bit more complicated because some missions have the vehicle land and disarm, but then be able to resume the mission from where they are in their sequence when they're armed in auto again.

Author: > And on re-upload, or manual reset to 0.

@julianoes Just to be clear, 
- I understand reupload on the ground will reset the counters and the current waypoint. 
- For upload in air the current waypoint does not change right? On PX4 you "finish the previously uploaded waypoint" and then the waypoint number increments and you progress to that waypoint in the newly uploaded mission.

> On manual reset to 0

What do you mean by this? Is this the same as ArduPilot - setting mission current to 0 resets the whole mission? If so, that would be good. 

> ArduPlane will not reset counters if you land the vehicle in auto. It's a bit more complicated because some missions have the vehicle land and disarm, but then be able to resume the mission from where they are in their sequence when they're armed in auto again.

@peterbarker Thanks very much. This caveat is true of all PX4 vehicle too - its a configuration setting or a mission type setting. I will add a comment.

Reviewer 2: > For upload in air the current waypoint does not change right? On PX4 you "finish the previously uploaded waypoint" and then the waypoint number increments and you progress to that waypoint in the newly uploaded mission.

This is ground station dependent. I think QGC does that. But if you just upload the whole mission, it will go back to 0.

> setting mission current to 0 resets the whole mission? If so, that would be good.

I think so, except things like DO_JUMP counters.

Author: @julianoes 

> I think so, except things like DO_JUMP counters.

So it will remove the "complete" bit and restart the mission, but with the wrong counters? That's screwed up. Can we make it also reset the counters? (i.e. as part of this protocol) and make not resetting them non-standard.

FYI also @peterbarker my concern here is that we updated [MAV_CMD_DO_SET_MISSION_CURRENT](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_SET_MISSION_CURRENT) to add a reset option because there was not way to reset the counters. Now it seems like setting the mission item to 0 would do this. That's a better approach. So we should standardize.
I.e. I think we'd be best off removing that separate parameter and stating that setting to zero should reset the counters and the complete bit and reset to zero. Thoughts?

Reviewer 2: > So it will remove the "complete" bit and restart the mission, but with the wrong counters? That's screwed up. Can we make it also reset the counters? (i.e. as part of this protocol) and make not resetting them non-standard.

That's the way I made it in 2013 or so. It might have been fixed since. You could try it out...

Author: I doubt it. In any case, Peter commented somewhere here he prefers that counters are reset explicitly. So we probably need to implement that second parameter. I looked at it once but couldn't work out the right place to even add the command.

For example PX4 will start as soon as you arm in mission mode, while ArduCopter in its default configuration requires that the throttle is raised.
Similarly, planes that are configured for throttle/hand launch may require a minimum acceleration before the mission is executed. 
While executing, the vehicle will progress sequentially through the mission items, looping and jumping in response to jump commands.

The mission can be paused by changing to another flight mode, such as Hold/loiter, and restarted by changing back to mission mode (when armed).
On restart the vehicle will _generally_ continue towards the same waypoint item as when it was paused.
In some cases flight stacks may start from the preceding waypoint, for example, a paused camera survey may restart on the previous waypoint to ensure the whole paused leg is captured.
Note that there is no concept of "stopping" a mission independent of pausing it.

On completion of all mission items the mission will "complete".
At this point some flight stacks will loiter within the mission mode, while others will transition to some other mode (such as "Hold") and not allow you to return to mission mode.

> In either case, the mission cannot be restarted unless it is [reset](#resetting-missions) (which usually happens after the vehicle lands and disarms).

Author: ```suggestion
In either case, the mission cannot be restarted unless it is [reset](#resetting-missions).
```
